File: ios/APEXNative/APEX/Core/Engine/FitnessEvidence.swift
import Foundation

enum FitnessEvidenceAdmission: String, Codable, Sendable, Equatable {
    case trusted
    case user
}

enum FitnessEvidenceConfidence: String, Codable, Sendable, Equatable, Hashable {
    case low
    case medium
    case high
}

enum FitnessEvidenceMetric: String, Codable, Sendable, Equatable, Hashable {
    case bodyMass = "body_mass"
    case height
    case bodyFatPercentage = "body_fat_percentage"
    case restingMetabolicRate = "resting_metabolic_rate"
    case vo2Max = "vo2_max"
    case restingHeartRate = "resting_heart_rate"
    case waistCircumference = "waist_circumference"
    case cardioCapacityScore = "cardio_capacity_score"
    case upperBodyStrengthScore = "upper_body_strength_score"
    case lowerBodyStrengthScore = "lower_body_strength_score"
    case flexibilityScore = "flexibility_score"
    case jointHealthScore = "joint_health_score"
    case balanceScore = "balance_score"
}

enum FitnessEvidenceSource: String, Codable, Sendable, Equatable, Hashable {
    case indirectCalorimetry = "indirect_calorimetry"
    case dexaMeasurement = "dexa_measurement"
    case dexaDerivedEstimate = "dexa_derived_estimate"
    case clinicalMeasurement = "clinical_measurement"
    case supportedDevice = "supported_device"
    case guidedAPEXFieldTest = "guided_apex_field_test"
    case structuredSelfReport = "structured_self_report"
    case userEnteredExternalResult = "user_entered_external_result"
    case legacyUnverified = "legacy_unverified"
}

struct FitnessEvidenceDraft: Codable, Sendable, Equatable {
    var userID: String
    var metric: String
    var value: Double
    var unit: String
    var source: String
    var `protocol`: String?
    var device: String?
    var measuredAt: String
    var importedAt: String
    var requestedConfidence: String
    var metadata: JSONValue
    var supersedesID: String?
    var clientIdempotencyKey: String

    enum CodingKeys: String, CodingKey {
        case userID = "user_id"
        case metric, value, unit, source, `protocol`, device, metadata
        case measuredAt = "measured_at"
        case importedAt = "imported_at"
        case requestedConfidence = "requested_confidence"
        case supersedesID = "supersedes_id"
        case clientIdempotencyKey = "client_idempotency_key"
    }
}

struct FitnessEvidencePredecessor: Codable, Sendable, Equatable {
    var id: String
    var userID: String
    var metric: String

    enum CodingKeys: String, CodingKey {
        case id, metric
        case userID = "user_id"
    }
}

struct NormalizedFitnessEvidence: Sendable, Equatable {
    var userID: String
    var metric: FitnessEvidenceMetric
    var value: Double
    var unit: String
    var source: FitnessEvidenceSource
    var `protocol`: String?
    var device: String?
    var measuredAt: String
    var importedAt: String
    var confidence: FitnessEvidenceConfidence
    var metadata: [String: JSONValue]
    var supersedesID: String?
    var clientIdempotencyKey: String
}

struct RecordUserFitnessEvidenceParameters: Encodable, Sendable {
    let pMetric: String
    let pValue: Double
    let pUnit: String
    let pMeasuredAt: String
    let pClientIdempotencyKey: String
    let pSource: String
    let pProtocol: String?
    let pDevice: String?
    let pMetadata: [String: JSONValue]
    let pSupersedesID: UUID?

    init(_ evidence: NormalizedFitnessEvidence) {
        pMetric = evidence.metric.rawValue
        pValue = evidence.value
        pUnit = evidence.unit
        pMeasuredAt = evidence.measuredAt
        pClientIdempotencyKey = evidence.clientIdempotencyKey
        pSource = evidence.source.rawValue
        pProtocol = evidence.protocol
        pDevice = evidence.device
        pMetadata = evidence.metadata
        pSupersedesID = evidence.supersedesID.flatMap(UUID.init(uuidString:))
    }

    enum CodingKeys: String, CodingKey {
        case pMetric = "p_metric"
        case pValue = "p_value"
        case pUnit = "p_unit"
        case pMeasuredAt = "p_measured_at"
        case pClientIdempotencyKey = "p_client_idempotency_key"
        case pSource = "p_source"
        case pProtocol = "p_protocol"
        case pDevice = "p_device"
        case pMetadata = "p_metadata"
        case pSupersedesID = "p_supersedes_id"
    }
}

enum FitnessEvidenceRecordingError: Error, Sendable {
    case accountMismatch
    case trustedSourceRequiresIngestion
}

struct FitnessEvidenceNormalizationSummary: Codable, Sendable, Equatable {
    var metric: FitnessEvidenceMetric
    var value: Double
    var unit: String
    var source: FitnessEvidenceSource
    var confidence: FitnessEvidenceConfidence
    var `protocol`: String?
    var device: String?
    var measuredAt: String
    var importedAt: String
    var metadataKeys: [String]
    var supersedesID: String?
    var clientIdempotencyKey: String

    enum CodingKeys: String, CodingKey {
        case metric, value, unit, source, confidence, `protocol`, device
        case measuredAt = "measured_at"
        case importedAt = "imported_at"
        case metadataKeys = "metadata_keys"
        case supersedesID = "supersedes_id"
        case clientIdempotencyKey = "client_idempotency_key"
    }
}

struct FitnessEvidenceNormalizationExpected: Codable, Sendable, Equatable {
    var status: String
    var summary: FitnessEvidenceNormalizationSummary?
    var reason: String?
}

struct FitnessEvidenceNormalizationFixture: Codable, Sendable {
    struct Scenario: Codable, Sendable {
        var name: String
        var admission: FitnessEvidenceAdmission
        var input: FitnessEvidenceDraft
        var predecessor: FitnessEvidencePredecessor?
        var expected: FitnessEvidenceNormalizationExpected
    }

    var referenceNow: String
    var scenarios: [Scenario]

    enum CodingKeys: String, CodingKey {
        case referenceNow = "reference_now"
        case scenarios
    }
}

enum FitnessEvidenceNormalizationResult: Sendable, Equatable {
    case accepted(NormalizedFitnessEvidence)
    case rejected(String)
}

enum FitnessEvidenceNormalizer {
    private static let metricAliases: [String: FitnessEvidenceMetric] = [
        "body_mass": .bodyMass,
        "body_weight": .bodyMass,
        "weight": .bodyMass,
        "weight_kg": .bodyMass,
        "height": .height,
        "height_cm": .height,
        "body_fat": .bodyFatPercentage,
        "body_fat_pct": .bodyFatPercentage,
        "body_fat_percentage": .bodyFatPercentage,
        "bmr": .restingMetabolicRate,
        "rmr": .restingMetabolicRate,
        "resting_metabolic_rate": .restingMetabolicRate,
        "vo2_max": .vo2Max,
        "vo2max": .vo2Max,
        "resting_heart_rate": .restingHeartRate,
        "resting_hr": .restingHeartRate,
        "waist": .waistCircumference,
        "waist_circumference": .waistCircumference,
        "cardio_capacity_score": .cardioCapacityScore,
        "upper_body_strength_score": .upperBodyStrengthScore,
        "lower_body_strength_score": .lowerBodyStrengthScore,
        "flexibility_score": .flexibilityScore,
        "joint_health_score": .jointHealthScore,
        "balance_score": .balanceScore
    ]

    private static let sourceAliases: [String: FitnessEvidenceSource] = [
        "indirect_calorimetry": .indirectCalorimetry,
        "metabolic_cart": .indirectCalorimetry,
        "dexa": .dexaMeasurement,
        "dexa_scan": .dexaMeasurement,
        "dexa_measurement": .dexaMeasurement,
        "dexa_derived": .dexaDerivedEstimate,
        "dexa_derived_estimate": .dexaDerivedEstimate,
        "clinical": .clinicalMeasurement,
        "clinical_measurement": .clinicalMeasurement,
        "supported_device": .supportedDevice,
        "apple_health": .supportedDevice,
        "apple_watch": .supportedDevice,
        "healthkit": .supportedDevice,
        "guided_apex_field_test": .guidedAPEXFieldTest,
        "apex_field_test": .guidedAPEXFieldTest,
        "structured_self_report": .structuredSelfReport,
        "self_report": .structuredSelfReport,
        "user_entered_external_result": .userEnteredExternalResult,
        "external_result": .userEnteredExternalResult,
        "legacy_unverified": .legacyUnverified,
        "legacy": .legacyUnverified
    ]

    private static let unitAliases: [String: String] = [
        "kg": "kg",
        "cm": "cm",
        "percent": "percent",
        "pct": "percent",
        "%": "percent",
        "kcal_per_day": "kcal_per_day",
        "kcal/day": "kcal_per_day",
        "ml_per_kg_min": "ml_per_kg_min",
        "ml/kg/min": "ml_per_kg_min",
        "bpm": "bpm",
        "beats_per_minute": "bpm",
        "score_0_100": "score_0_100"
    ]

    private static let trustedSources: Set<FitnessEvidenceSource> = [
        .indirectCalorimetry,
        .dexaMeasurement,
        .dexaDerivedEstimate,
        .clinicalMeasurement,
        .supportedDevice,
        .guidedAPEXFieldTest
    ]

    private static let confidenceRank: [FitnessEvidenceConfidence: Int] = [
        .low: 0,
        .medium: 1,
        .high: 2
    ]

    private static let confidenceCeiling: [FitnessEvidenceSource: FitnessEvidenceConfidence] = [
        .indirectCalorimetry: .high,
        .dexaMeasurement: .high,
        .dexaDerivedEstimate: .medium,
        .clinicalMeasurement: .high,
        .supportedDevice: .medium,
        .guidedAPEXFieldTest: .medium,
        .structuredSelfReport: .low,
        .userEnteredExternalResult: .low,
        .legacyUnverified: .low
    ]

    private enum OptionalTextResult {
        case valid(String?)
        case invalid
    }

    static func normalize(
        _ input: FitnessEvidenceDraft,
        admission: FitnessEvidenceAdmission,
        referenceNow: String,
        predecessor: FitnessEvidencePredecessor? = nil
    ) -> FitnessEvidenceNormalizationResult {
        let userID = input.userID.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !userID.isEmpty else { return .rejected("invalid_owner") }

        guard let metric = metricAliases[token(input.metric)] else {
            return .rejected("unsupported_metric")
        }
        guard let submittedSource = sourceAliases[token(input.source)] else {
            return .rejected("unsupported_source")
        }
        guard let unit = normalizeUnit(input.unit), valid(metric: metric, unit: unit, value: input.value) else {
            return .rejected("invalid_unit_or_range")
        }
        guard case .object(let metadata) = input.metadata else {
            return .rejected("invalid_metadata")
        }

        guard
            let measured = parseDate(input.measuredAt),
            let imported = parseDate(input.importedAt),
            let reference = parseDate(referenceNow)
        else {
            return .rejected("invalid_timestamp")
        }
        let oneDay: TimeInterval = 24 * 60 * 60
        let earliest = parseDate("1900-01-01T00:00:00Z") ?? .distantPast
        guard
            measured >= earliest,
            measured <= imported.addingTimeInterval(oneDay),
            imported <= reference.addingTimeInterval(oneDay)
        else {
            return .rejected("invalid_timestamp")
        }

        let clientKey = input.clientIdempotencyKey.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !clientKey.isEmpty, clientKey.count <= 160 else {
            return .rejected("invalid_idempotency_key")
        }

        let protocolResult = optionalText(input.protocol, maximum: 160)
        let deviceResult = optionalText(input.device, maximum: 200)
        guard
            case .valid(let submittedProtocol) = protocolResult,
            case .valid(let device) = deviceResult
        else {
            return .rejected("invalid_text_field")
        }

        var source = submittedSource
        var normalizedProtocol = submittedProtocol
        if admission == .user, trustedSources.contains(submittedSource) {
            source = .userEnteredExternalResult
            normalizedProtocol = normalizedProtocol ?? "reported:\(submittedSource.rawValue)"
        } else if admission == .user, submittedSource == .legacyUnverified {
            source = .userEnteredExternalResult
            normalizedProtocol = normalizedProtocol ?? "reported:legacy_unverified"
        }

        guard let requested = FitnessEvidenceConfidence(rawValue: input.requestedConfidence),
              let ceiling = confidenceCeiling[source],
              let requestedRank = confidenceRank[requested],
              let ceilingRank = confidenceRank[ceiling]
        else {
            return .rejected("invalid_confidence")
        }
        let confidence = requestedRank <= ceilingRank ? requested : ceiling

        let supersedesID = input.supersedesID?
            .trimmingCharacters(in: .whitespacesAndNewlines)
            .nilIfEmpty
        if let supersedesID {
            let predecessorMetric = predecessor.flatMap { metricAliases[token($0.metric)] }
            guard
                let predecessor,
                predecessor.id == supersedesID,
                predecessor.userID == userID,
                predecessorMetric == metric
            else {
                return .rejected("invalid_correction")
            }
        }

        return .accepted(NormalizedFitnessEvidence(
            userID: userID,
            metric: metric,
            value: input.value,
            unit: unit,
            source: source,
            protocol: normalizedProtocol,
            device: device,
            measuredAt: formatDate(measured),
            importedAt: formatDate(imported),
            confidence: confidence,
            metadata: metadata,
            supersedesID: supersedesID,
            clientIdempotencyKey: clientKey
        ))
    }

    static func summarize(_ result: FitnessEvidenceNormalizationResult) -> FitnessEvidenceNormalizationExpected {
        switch result {
        case .rejected(let reason):
            return FitnessEvidenceNormalizationExpected(status: "rejected", summary: nil, reason: reason)
        case .accepted(let evidence):
            return FitnessEvidenceNormalizationExpected(
                status: "accepted",
                summary: FitnessEvidenceNormalizationSummary(
                    metric: evidence.metric,
                    value: evidence.value,
                    unit: evidence.unit,
                    source: evidence.source,
                    confidence: evidence.confidence,
                    protocol: evidence.protocol,
                    device: evidence.device,
                    measuredAt: evidence.measuredAt,
                    importedAt: evidence.importedAt,
                    metadataKeys: evidence.metadata.keys.sorted(),
                    supersedesID: evidence.supersedesID,
                    clientIdempotencyKey: evidence.clientIdempotencyKey
                ),
                reason: nil
            )
        }
    }

    private static func token(_ value: String) -> String {
        let lowered = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        var result = ""
        var previousWasSeparator = false
        for scalar in lowered.unicodeScalars {
            if CharacterSet.alphanumerics.contains(scalar) || scalar.value == 37 {
                result.unicodeScalars.append(scalar)
                previousWasSeparator = false
            } else if !previousWasSeparator, !result.isEmpty {
                result.append("_")
                previousWasSeparator = true
            }
        }
        return result.trimmingCharacters(in: CharacterSet(charactersIn: "_"))
    }

    private static func normalizeUnit(_ value: String) -> String? {
        let raw = value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        return unitAliases[raw] ?? unitAliases[token(raw)]
    }

    private static func valid(metric: FitnessEvidenceMetric, unit: String, value: Double) -> Bool {
        guard value.isFinite else { return false }
        switch metric {
        case .bodyMass:
            return unit == "kg" && (10...500).contains(value)
        case .height:
            return unit == "cm" && (50...260).contains(value)
        case .bodyFatPercentage:
            return unit == "percent" && (2...70).contains(value)
        case .restingMetabolicRate:
            return unit == "kcal_per_day" && (400...8000).contains(value)
        case .vo2Max:
            return unit == "ml_per_kg_min" && (5...120).contains(value)
        case .restingHeartRate:
            return unit == "bpm" && (20...250).contains(value)
        case .waistCircumference:
            return unit == "cm" && (30...300).contains(value)
        case .cardioCapacityScore,
             .upperBodyStrengthScore,
             .lowerBodyStrengthScore,
             .flexibilityScore,
             .jointHealthScore,
             .balanceScore:
            return unit == "score_0_100" && (0...100).contains(value)
        }
    }

    private static func optionalText(_ value: String?, maximum: Int) -> OptionalTextResult {
        guard let value else { return .valid(nil) }
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return .valid(nil) }
        guard trimmed.count <= maximum else { return .invalid }
        return .valid(trimmed)
    }

    private static func parseDate(_ value: String) -> Date? {
        let fractional = ISO8601DateFormatter()
        fractional.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = fractional.date(from: value) { return date }
        let wholeSeconds = ISO8601DateFormatter()
        wholeSeconds.formatOptions = [.withInternetDateTime]
        return wholeSeconds.date(from: value)
    }

    private static func formatDate(_ value: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        formatter.timeZone = TimeZone(secondsFromGMT: 0)
        return formatter.string(from: value)
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: src/lib/fitnessEvidence.ts
export type FitnessEvidenceAdmission = 'trusted' | 'user'
export type FitnessEvidenceConfidence = 'low' | 'medium' | 'high'

export type FitnessEvidenceMetric =
  | 'body_mass'
  | 'height'
  | 'body_fat_percentage'
  | 'resting_metabolic_rate'
  | 'vo2_max'
  | 'resting_heart_rate'
  | 'waist_circumference'
  | 'cardio_capacity_score'
  | 'upper_body_strength_score'
  | 'lower_body_strength_score'
  | 'flexibility_score'
  | 'joint_health_score'
  | 'balance_score'

export type FitnessEvidenceSource =
  | 'indirect_calorimetry'
  | 'dexa_measurement'
  | 'dexa_derived_estimate'
  | 'clinical_measurement'
  | 'supported_device'
  | 'guided_apex_field_test'
  | 'structured_self_report'
  | 'user_entered_external_result'
  | 'legacy_unverified'

export interface FitnessEvidenceDraft {
  user_id: string
  metric: string
  value: number
  unit: string
  source: string
  protocol?: string | null
  device?: string | null
  measured_at: string
  imported_at: string
  requested_confidence: string
  metadata: unknown
  supersedes_id?: string | null
  client_idempotency_key: string
}
export interface FitnessEvidencePredecessor {
  id: string
  user_id: string
  metric: string
}

export interface NormalizedFitnessEvidence {
  user_id: string
  metric: FitnessEvidenceMetric
  value: number
  unit: string
  source: FitnessEvidenceSource
  protocol: string | null
  device: string | null
  measured_at: string
  imported_at: string
  confidence: FitnessEvidenceConfidence
  metadata: Record<string, unknown>
  supersedes_id: string | null
  client_idempotency_key: string
}

export type FitnessEvidenceRejection =
  | 'invalid_owner'
  | 'unsupported_metric'
  | 'unsupported_source'
  | 'invalid_unit_or_range'
  | 'invalid_timestamp'
  | 'invalid_metadata'
  | 'invalid_idempotency_key'
  | 'invalid_confidence'
  | 'invalid_correction'
  | 'invalid_text_field'

export type FitnessEvidenceNormalizationResult =
  | { status: 'accepted'; evidence: NormalizedFitnessEvidence }
  | { status: 'rejected'; reason: FitnessEvidenceRejection }

export interface FitnessEvidenceNormalizationSummary {
  status: 'accepted'
  summary: {
    metric: FitnessEvidenceMetric
    value: number
    unit: string
    source: FitnessEvidenceSource
    confidence: FitnessEvidenceConfidence
    protocol: string | null
    device: string | null
    measured_at: string
    imported_at: string
    metadata_keys: string[]
    supersedes_id: string | null
    client_idempotency_key: string
  }
}

export interface FitnessEvidenceNormalizationScenario {
  name: string
  admission: FitnessEvidenceAdmission
  input: FitnessEvidenceDraft
  predecessor?: FitnessEvidencePredecessor
  expected:
    | FitnessEvidenceNormalizationSummary
    | { status: 'rejected'; reason: FitnessEvidenceRejection }
}

export interface FitnessEvidenceNormalizationFixture {
  reference_now: string
  scenarios: FitnessEvidenceNormalizationScenario[]
}

const metricAliases: Readonly<Record<string, FitnessEvidenceMetric>> = {
  body_mass: 'body_mass',
  body_weight: 'body_mass',
  weight: 'body_mass',
  weight_kg: 'body_mass',
  height: 'height',
  height_cm: 'height',
  body_fat: 'body_fat_percentage',
  body_fat_pct: 'body_fat_percentage',
  body_fat_percentage: 'body_fat_percentage',
  bmr: 'resting_metabolic_rate',
  rmr: 'resting_metabolic_rate',
  resting_metabolic_rate: 'resting_metabolic_rate',
  vo2_max: 'vo2_max',
  vo2max: 'vo2_max',
  resting_heart_rate: 'resting_heart_rate',
  resting_hr: 'resting_heart_rate',
  waist: 'waist_circumference',
  waist_circumference: 'waist_circumference',
  cardio_capacity_score: 'cardio_capacity_score',
  upper_body_strength_score: 'upper_body_strength_score',
  lower_body_strength_score: 'lower_body_strength_score',
  flexibility_score: 'flexibility_score',
  joint_health_score: 'joint_health_score',
  balance_score: 'balance_score',
}

const sourceAliases: Readonly<Record<string, FitnessEvidenceSource>> = {
  indirect_calorimetry: 'indirect_calorimetry',
  metabolic_cart: 'indirect_calorimetry',
  dexa: 'dexa_measurement',
  dexa_scan: 'dexa_measurement',
  dexa_measurement: 'dexa_measurement',
  dexa_derived: 'dexa_derived_estimate',
  dexa_derived_estimate: 'dexa_derived_estimate',
  clinical: 'clinical_measurement',
  clinical_measurement: 'clinical_measurement',
  supported_device: 'supported_device',
  apple_health: 'supported_device',
  apple_watch: 'supported_device',
  healthkit: 'supported_device',
  guided_apex_field_test: 'guided_apex_field_test',
  apex_field_test: 'guided_apex_field_test',
  structured_self_report: 'structured_self_report',
  self_report: 'structured_self_report',
  user_entered_external_result: 'user_entered_external_result',
  external_result: 'user_entered_external_result',
  legacy_unverified: 'legacy_unverified',
  legacy: 'legacy_unverified',
}

const unitAliases: Readonly<Record<string, string>> = {
  kg: 'kg',
  cm: 'cm',
  percent: 'percent',
  pct: 'percent',
  '%': 'percent',
  kcal_per_day: 'kcal_per_day',
  'kcal/day': 'kcal_per_day',
  ml_per_kg_min: 'ml_per_kg_min',
  'ml/kg/min': 'ml_per_kg_min',
  bpm: 'bpm',
  beats_per_minute: 'bpm',
  score_0_100: 'score_0_100',
}

const trustedSources = new Set<FitnessEvidenceSource>([
  'indirect_calorimetry',
  'dexa_measurement',
  'dexa_derived_estimate',
  'clinical_measurement',
  'supported_device',
  'guided_apex_field_test',
])

const confidenceRank: Readonly<Record<FitnessEvidenceConfidence, number>> = {
  low: 0,
  medium: 1,
  high: 2,
}

const confidenceCeiling: Readonly<Record<FitnessEvidenceSource, FitnessEvidenceConfidence>> = {
  indirect_calorimetry: 'high',
  dexa_measurement: 'high',
  dexa_derived_estimate: 'medium',
  clinical_measurement: 'high',
  supported_device: 'medium',
  guided_apex_field_test: 'medium',
  structured_self_report: 'low',
  user_entered_external_result: 'low',
  legacy_unverified: 'low',
}

function token(value: string): string {
  return value.trim().toLowerCase().replace(/[^a-z0-9%]+/g, '_').replace(/^_+|_+$/g, '')
}

function normalizeMetric(value: string): FitnessEvidenceMetric | null {
  return metricAliases[token(value)] ?? null
}

function normalizeSource(value: string): FitnessEvidenceSource | null {
  return sourceAliases[token(value)] ?? null
}

function normalizeUnit(value: string): string | null {
  const raw = value.trim().toLowerCase()
  return unitAliases[raw] ?? unitAliases[token(raw)] ?? null
}

function validMetricUnitRange(metric: FitnessEvidenceMetric, unit: string, value: number): boolean {
  if (!Number.isFinite(value)) return false
  switch (metric) {
    case 'body_mass': return unit === 'kg' && value >= 10 && value <= 500
    case 'height': return unit === 'cm' && value >= 50 && value <= 260
    case 'body_fat_percentage': return unit === 'percent' && value >= 2 && value <= 70
    case 'resting_metabolic_rate': return unit === 'kcal_per_day' && value >= 400 && value <= 8000
    case 'vo2_max': return unit === 'ml_per_kg_min' && value >= 5 && value <= 120
    case 'resting_heart_rate': return unit === 'bpm' && value >= 20 && value <= 250
    case 'waist_circumference': return unit === 'cm' && value >= 30 && value <= 300
    case 'cardio_capacity_score':
    case 'upper_body_strength_score':
    case 'lower_body_strength_score':
    case 'flexibility_score':
    case 'joint_health_score':
    case 'balance_score':
      return unit === 'score_0_100' && value >= 0 && value <= 100
  }
}

function normalizedDate(value: string): { date: Date; iso: string } | null {
  const date = new Date(value)
  if (!Number.isFinite(date.getTime())) return null
  return { date, iso: date.toISOString() }
}

function optionalText(value: string | null | undefined, maximum: number): string | null | undefined {
  if (value == null) return null
  const result = value.trim()
  if (!result) return null
  return result.length <= maximum ? result : undefined
}

function cappedConfidence(
  requested: string,
  source: FitnessEvidenceSource,
): FitnessEvidenceConfidence | null {
  if (requested !== 'low' && requested !== 'medium' && requested !== 'high') return null
  const ceiling = confidenceCeiling[source]
  return confidenceRank[requested] <= confidenceRank[ceiling] ? requested : ceiling
}

function rejected(reason: FitnessEvidenceRejection): FitnessEvidenceNormalizationResult {
  return { status: 'rejected', reason }
}

export function normalizeFitnessEvidence(
  input: FitnessEvidenceDraft,
  admission: FitnessEvidenceAdmission,
  referenceNow: string,
  predecessor: FitnessEvidencePredecessor | null = null,
): FitnessEvidenceNormalizationResult {
  const userID = input.user_id.trim()
  if (!userID) return rejected('invalid_owner')

  const metric = normalizeMetric(input.metric)
  if (!metric) return rejected('unsupported_metric')

  const submittedSource = normalizeSource(input.source)
  if (!submittedSource) return rejected('unsupported_source')

  const unit = normalizeUnit(input.unit)
  if (!unit || !validMetricUnitRange(metric, unit, input.value)) {
    return rejected('invalid_unit_or_range')
  }

  if (input.metadata == null || typeof input.metadata !== 'object' || Array.isArray(input.metadata)) {
    return rejected('invalid_metadata')
  }

  const measured = normalizedDate(input.measured_at)
  const imported = normalizedDate(input.imported_at)
  const reference = normalizedDate(referenceNow)
  if (!measured || !imported || !reference) return rejected('invalid_timestamp')
  const oneDay = 24 * 60 * 60 * 1000
  const earliest = Date.parse('1900-01-01T00:00:00Z')
  if (
    measured.date.getTime() < earliest
    || measured.date.getTime() > imported.date.getTime() + oneDay
    || imported.date.getTime() > reference.date.getTime() + oneDay
  ) {
    return rejected('invalid_timestamp')
  }

  const clientKey = input.client_idempotency_key.trim()
  if (!clientKey || clientKey.length > 160) return rejected('invalid_idempotency_key')

  const normalizedProtocol = optionalText(input.protocol, 160)
  const normalizedDevice = optionalText(input.device, 200)
  if (normalizedProtocol === undefined || normalizedDevice === undefined) {
    return rejected('invalid_text_field')
  }

  let source = submittedSource
  let protocol = normalizedProtocol
  if (admission === 'user' && trustedSources.has(submittedSource)) {
    source = 'user_entered_external_result'
    protocol = protocol ?? `reported:${submittedSource}`
  } else if (admission === 'user' && submittedSource === 'legacy_unverified') {
    source = 'user_entered_external_result'
    protocol = protocol ?? 'reported:legacy_unverified'
  }

  const confidence = cappedConfidence(input.requested_confidence, source)
  if (!confidence) return rejected('invalid_confidence')

  const supersedesID = input.supersedes_id?.trim() || null
  if (supersedesID) {
    const predecessorMetric = predecessor ? normalizeMetric(predecessor.metric) : null
    if (
      !predecessor
      || predecessor.id !== supersedesID
      || predecessor.user_id !== userID
      || predecessorMetric !== metric
    ) {
      return rejected('invalid_correction')
    }
  }

  return {
    status: 'accepted',
    evidence: {
      user_id: userID,
      metric,
      value: input.value,
      unit,
      source,
      protocol,
      device: normalizedDevice,
      measured_at: measured.iso,
      imported_at: imported.iso,
      confidence,
      metadata: { ...(input.metadata as Record<string, unknown>) },
      supersedes_id: supersedesID,
      client_idempotency_key: clientKey,
    },
  }
}

export function summarizeFitnessEvidenceNormalization(
  result: FitnessEvidenceNormalizationResult,
): FitnessEvidenceNormalizationSummary | { status: 'rejected'; reason: FitnessEvidenceRejection } {
  if (result.status === 'rejected') return result
  const evidence = result.evidence
  return {
    status: 'accepted',
    summary: {
      metric: evidence.metric,
      value: evidence.value,
      unit: evidence.unit,
      source: evidence.source,
      confidence: evidence.confidence,
      protocol: evidence.protocol,
      device: evidence.device,
      measured_at: evidence.measured_at,
      imported_at: evidence.imported_at,
      metadata_keys: Object.keys(evidence.metadata).sort(),
      supersedes_id: evidence.supersedes_id,
      client_idempotency_key: evidence.client_idempotency_key,
    },
  }
}
